File: ts/app/http/request.ts
import net from 'net'

import { Response } from './response'
import { encodingSelection } from '../encode'

const crlf = '\r\n'

export class Request {
    method: string  // GET/POST currently supported
    url: string
    path: string[]
    version: string // Only HTTP 1.1 implemented

    private header: string
    headers: { [header: string]: string | string[] } // Not many supported yet
    body: string    // Only strings supported currently

    constructor(data: Buffer, private socket: net.Socket) {
        // Title
        [this.header, this.body] = data.toString().split(crlf + crlf)
        const lines = this.header.toString().split(crlf);
        [this.method, this.url, this.version] = lines[0].split(' ')
        // this.method = lineOne[0], this.url = lineOne[1],
        this.path = this.url.split('/').filter(val => val.length > 0) // Sanitize path from empty sections

        // Headers
        const headers = lines.slice(1).map(line => line.split(': '))
        this.headers = {}
        for (const [key, value] of headers)
            if (this.headers[key])  // Value exists, convert/insert to array
                if (Array.isArray(this.headers[key])) this.headers[key].push(value)
                else this.headers[key] = [this.headers[key], value]
            else this.headers[key] = value
    }

    respond(code: number | string, payload) {
        console.log('->', this.socket.remoteAddress, code, payload)
        const response = new Response(code, payload)

        const encoding = encodingSelection(this.headers['Accept-Encoding'])
        if (encoding) response.encoding = encoding

        const headers = response.headers(), body = response.payloadBuffer()
        console.log('  ', response.contentType, response.length(), 'bytes')

        this.socket.write(headers)
        this.socket.write(new Uint8Array(body))
        this.socket.end()
    }
}

File: src/http/request.cpp
#include "request.hpp"

#include <arpa/inet.h>
#include <sys/socket.h>
#include <vector>

#include "../utils/console.hpp"
#include "../utils/str.hpp"

#define BUFFER_SIZE 1024

using std::string;
using std::vector;

Request::Request() {}

string Request::ip() {
    //            num to addr                   .s_addr = IP as binary (32b)
    return string(inet_ntoa(client_addr.sin_addr));
}

void Request::accept(int server_fd) {
    socklen_t len = sizeof(client_addr); // Must be declared in memory for there to be a pointer to give ::accept()
    //          global-scoped (from socket.h)
    client_fd = ::accept(server_fd, (struct sockaddr *)&client_addr, &len);

    // Read contents
    char buffer[BUFFER_SIZE] = {0}; // 1MB
    if (recv(client_fd, buffer, BUFFER_SIZE, 0) < 0)
        throw "Message receival failed";
    contents = string(buffer);

    // Analyze header
    string header = split(contents, "\r\n\r\n")[0];
    vector<string> headerLines = split(header, "\r\n");
    vector<string> title = split(headerLines[0], " ");
    method = title[0];
    path = title[1];
    pathComponents = filter(split(path, "/"), "");
    for (int i = 1; i < headerLines.size(); i++) {
        const vector<string> line = split(headerLines[i], ": ");
        if (line.size() < 2) continue; // skip malformed lines

        const string key = line[0];
        const vector<string> values = { line[1] };

        for (const auto &value : values) headers[key].push_back(value);
    }

    // Log
    log(ip() + " -> " + method + " " + path);
}

void Request::respond(int code, string message) { respond(new Response(code, message), 0); }
void Request::respond(Response *response, int options) {
    string headers = response->headers();
    string msg = response->getMessage();
    log(std::to_string(response->getCode()) + " -> " + ip() + (msg.empty() ? "" : ":"));
    if (!msg.empty())
        log(msg);
    //              Returns a char* from the str
    send(client_fd, headers.c_str(), headers.size(), options);
    send(client_fd, msg.c_str(), msg.size(), options);
}

vector<string> Request::getPath() { return pathComponents; }
string Request::fullPath() { return path; }

vector<string> Request::getHeader(string key) { return headers[key]; }